tensorboard/data/server/reservoir.rs

```diff
@@ -0,0 +1,329 @@
+    /// generator for reservoir control. All reservoirs created by this function will use the same
+    /// sequence of random numbers.
+    ///
+    /// This function does not allocate. Reservoir capacity is allocated as records are offered.
+    pub fn new(capacity: usize) -> Self {
+        Self::with_control(capacity, ChaCha20Rng::seed_from_u64(0))
+    }
+}
+
+impl<T, C: ReservoirControl> StageReservoir<T, C> {
+    /// Offers a record to the reservoir. The reservoir will always include the latest record.
+    /// Other than the latest record, the records kept form a simple random sample of the stream
+    /// (or at least approximately so in the case of preemptions).
+    pub fn offer(&mut self, step: Step, v: T) {
+        self.seen += 1;
+        let dst = self.ctl.destination(self.seen);
+
+        // Didn't make the cut? Keep-last only.
+        if dst >= self.capacity {
+            if self.staged_items.pop().is_none() {
+                self.committed_steps.pop();
+            }
+            self.staged_items.push((step, v));
+            return;
+        }
+
+        // Room to spare? Add to end.
+        let num_stored = self.committed_steps.len() + self.staged_items.len();
+        if num_stored < self.capacity {
+            self.staged_items.push((step, v));
+            return;
+        }
+
+        // Otherwise, we're evicting.
+        if dst < self.committed_steps.len() {
+            self.committed_steps.remove(dst);
+        } else {
+            self.staged_items.remove(dst - self.committed_steps.len());
+        }
+        self.staged_items.push((step, v))
+    }
+
+    /// Creates a new reservoir with the specified capacity and reservoir control.
+    ///
+    /// This function does not allocate. Reservoir capacity is allocated as records are offered.
+    pub fn with_control(capacity: usize, ctl: C) -> Self {
+        Self {
+            committed_steps: Vec::new(),
+            staged_items: Vec::new(),
+            capacity,
+            ctl,
+            seen: 0,
+        }
+    }
+
+    /// Accesses a view of the currently staged items.
+    pub fn staged_items(&self) -> &[(Step, T)] {
+        &self.staged_items[..]
+    }
+
+    /// Commits pending changes from this reservoir into a commit view. The commit should be a
+    /// vector that starts empty and is modified only by calls to `commit`/`commit_map` on this
+    /// reservoir value.
+    pub fn commit(&mut self, head: &mut Vec<(Step, T)>) {
+        self.commit_map(head, |t| t)
+    }
+
+    /// Commits pending changes from this reservoir into a commit view, applying a mapping function
+    /// to each new value. This can be used to perform relatively expensive conversions or
+    /// enrichments only for records that are actually committed. The commit should be a vector
+    /// that starts empty and is modified only by calls to `commit`/`commit_map` on this reservoir
```

Comment:
Yep, you could certainly do that—which would buy you the “only modified
by commits on _a_ reservoir” part, but not “[…] on _this_ reservoir”.
Getting the latter without dependent types looks a bit tricky. You can
kind of get it with an `Arc` penalty—

```rust
impl StageReservoir {
    pub fn new() -> (Self, Arc<Commit>);
}
```

—which still lets you access private data/methods, but you won’t have a
`&mut` reference. And you could move the `RwLock` in here, but we
weren’t planning to `RwLock` around every reservoir.

Anyway, yeah, you can do what you said. I didn’t do so because I didn’t
think it was necessary and I wanted to keep things simple. How much is
this “let’s do this” versus just exploring the possibilities of the type
system?

Reply:
(Done via `Basin<T>`; see above.)

Reply:
(Yeah, I left the comment in there for the “same reservoir” part.)